### ISSUE TITLE
[BUG] "Travel rules not up to date" / "Reiseregeln nicht aktuell": Unfortunately the last update failed…

### ISSUE THREAD

**Issue by Jo-Achim** [Digitaler-Impfnachweis/covpass-android#98]
## Avoid duplicates
* [X] Bug is not mentioned in the [FAQ](https://digitaler-impfnachweis-app.de/faq/)
* [X] Bug is not already reported in another issue
* [X] OS: Android. _iOS not affected; see: https://github.com/Digitaler-Impfnachweis/covpass-android/issues/98#issuecomment-960642531._ 

## Technical details

- Device name: Not relevant. (CovPass v1.12.0: Samsung Galaxy Note 10, SM-N970F/DS  ///  CovPass v 1.11.1: Nokia 7.2, TA-1196)
- OS version: Not relevant. (Samsung: Android 11 (One UI 3.1) with Android security update: 01. Oktober 2021  ///  Nokia: Android 10 with Android security update: September 1st, 2021)
- App version: 1.12.0

## Describe the bug

When calling **CovPass v1.12.0** and calling "Check validity >", the following message is displayed in the "Check validity" dialog:

"Travel rules not up to date" ...
![Screenshot_20211103-232151_CovPass_2](https://user-images.githubusercontent.com/21124877/140298775-6328a353-c8f5-4d69-babd-52298bf71d5e.jpg)

It seems to me that the app makes no attempt to update, because the notice is displayed immediately, despite the presence of WiFi.

In comparison to the previous **version 1.11.1**, the following is initially displayed if the above procedure is the same:
![Screenshot_20211104-004356_2](https://user-images.githubusercontent.com/21124877/140292553-1d1709e9-aebd-4766-a845-e85b2b6e6793.jpg)

and then a few seconds later:
![Screenshot_20211104-004409_2](https://user-images.githubusercontent.com/21124877/140292665-67b65025-92e3-47ff-8f0b-47567c94429c.jpg)

## Steps to reproduce the issue

Make sure that the smartphone has an internet connection - in my case **only via WLAN**! (One smartphone (Samsung) is not allowed to access the internet via mobile phone provider, the second one (Nokia) does not have a SIM card.)

1. Start CovPass v1.12.0
2. Tap on "Check Validity >"
3. The v1.12.0 dialog shown above is displayed - without any time delay. 

## Expected behaviour

Ensuring that every time CovPass v1.12.0 is started, the 'out of date data for the entry rules' are updated via the existing Internet.

Cross reference: [["CovPass Check": Offline mode & WLAN]: "Internet connection required" although available ...](https://github.com/Digitaler-Impfnachweis/documentation/issues/22#issue-950487892).
(Please do not be surprised about the version numbers of CovPass / CovPassCheck mentioned in the cross-reference, which are 'normal'; see [Repeated version jumps / incomprehensible numbering of the versions in CovPass & CovPassCheck ...](https://github.com/Digitaler-Impfnachweis/covpass-android/issues/92#issue-1033356078).) 

## Possible Fix

Implementation of the update mechanism for the travel rules from version 1.11.1 or other corrections in the next version of CovPass.

## Additional context

Open / not tested: similar behavior in CovPassCheck v 1.12.0. 

-------------

PS: If helpful, a supplement for the above Samsung data:

CovPassCheck v.1.12.0 shows:
- Offline mode available
- List of certificate issuers: 11/03/2021 6:09 pm
- Inspection rules: 11/02/2021 6:15 pm.

**Comment by Ein-Tim** [comment]
Can't reproduce this under iOS.

**Comment by Jo-Achim** [comment]
**Update:** Since yesterday afternoon (11/05/2021), round about 36 hours after update to version 1.12.0, the message "Travel rules not up to date" / "Reiseregeln nicht aktuell" has disappeared without further ado.

Cross reference: [CovPassCheck shows warning "Check date and time"](https://github.com/Digitaler-Impfnachweis/covpass-android/issues/97#issue-1043211492).

Something is not going well.

**Comment by Jo-Achim** [comment]
Info:
"Travel rules not up to date" / "Reiseregeln nicht aktuell" again not up to date; 11/08/2021 ~0:0x am (or earlier).
11/08/2021 ~0:08 a.m., no more corresponding message.

**Comment by Jo-Achim** [comment]
Today (11/13/2021 06:38 pm): 
"Travel rules not up to date" / "Reiseregeln nicht aktuell" again. 

Something goes wrong!

See cross reference from today: https://github.com/Digitaler-Impfnachweis/covpass-android/issues/97#issuecomment-968069767!

**Comment by MikeMcC399** [comment]
@Jo-Achim 
It sounds like you have connectivity problems from your device, especially if there are not any similar reports from other users.

Are you running any firewall software on the device or are you behind a firewall?

**Comment by Jo-Achim** [comment]
@MikeMcC399,

yes, there is a firewall involved; [NetGuard](https://github.com/M66B/NetGuard) v2.298.

However, I would like to point out that ...
- Changes in handling between CovPass 1.12.0 and CovPass 1.11.1 have been made; see first (v1.12.0) and second (v1.11.1) screenshot above.
- Similar problems have already been 'solved' in CovPassCheck with the updates from v1.28.4 / 1.72.4 to v1.95.12. See "cross reference" above ([["CovPass Check": Offline mode & WLAN]: "Internet connection required" although available ...](https://github.com/Digitaler-Impfnachweis/documentation/issues/22#issuecomment-917366087)).

The same firewall with the same configuration is also installed on my test / comparison cell phone. So that I can say in comparison: Version 1.11.1 (red frame) and 1.12.0 behave differently under the same configurations. And since I suspect very similar mechanisms in CovPass and CovPassCheck...

Finally: CWA has **not** had any problems since then; see [[INFO] URSACHE: 9002, Etwas ist schiefgelaufen. Timeout](https://github.com/corona-warn-app/cwa-app-android/issues/998#issuecomment-674369983) #998).

**Comment by JilAuraya** [comment]
Hello Jo-Achim, thank you for pointing out this bug (and I apologize for the late response). 
This will be fixed with the upcoming release of the CovPass app.

**Comment by Jo-Achim** [comment]
@JilAuraya, thank you.

**Comment by Jo-Achim** [comment]
_Additional information:_
Tonight the message "Travel rules not up to date" again.
Then I deleted the cache (via Settings, Apps, CovPass) from CovPass ... but the message stayed after restarting the app.

Then I deleted the cache again plus the data from CovPass. The result is hardly surprising: after all vaccination certificates were scanned again, the message was no longer available.

PS: WLAN was permanently available. According to the NetGuard protocol, there were no 'blocked' access attempts.

**Comment by timokoenig** [comment]
As far as I can see, the worker to update the rules runs every 24 hours in the background. The message "Travel rules not up to date" gets displayed when the rules are older than 24 hours. Even when the requests do not get blocked by a firewall there is a small window where the message will appear due to the time it takes to update the rules. If there is an error and the worker goes into the exponential backoff this window might goes from a few minutes to a few hours.

@JilAuraya @pucherd I suggest to change the time to show the message to 36 or 48 hours. This way there is small buffer where the app has time to update the rules without suggesting that the app is out of date.

Another addition would be to show the actual error code when the background worker fails so that the user has a better indication whether something is wrong or not

**Comment by pucherd** [comment]
Hi @timokoenig, we replaced all workers with "on-demand" requests (v1.13.1) which will be done on users interaction -> so each time the user lands in the card overview we will check if the rules are older than 24 hours.
But it is a good Idea to show the actual error code when the request fails.

**Comment by timokoenig** [comment]
@pucherd that's great to hear! I guess that will mitigate some issues with the background tasks

**Comment by Jo-Achim** [comment]
Thank you, @pucherd 
> ..., we replaced all workers with "on-demand" requests (v1.13.1) which will be done on users interaction -> so each time the user lands in the card overview we will check if the rules are older than 24 hours. But it is a good Idea to show the actual error code when the request fails.

With the current version 1.13.2 the optional manual option "Update entry rules" is available.

**Comment by pucherd** [comment]
Thank you for the feedback @Jo-Achim. Can we close this ticket?

**Comment by Jo-Achim** [comment]
Yes, I close, since the alternative implementation via the manual update is also a good solution.
The other solution would of course have been, if there was any internet connection, to update the travel rules automatically every time the app was started. As probably in version 1.11.1.

**Comment by pucherd** [comment]
@Jo-Achim the manual update is a alternative solution the travel rules will be updated each time the app is started (of course if the the last update is older than 24 hours).
